File: dotnet/src/Oppex.Integration.Sdk/IncidentRequest.cs
namespace Oppex.Integration.Sdk;

/// <summary>A single incident submission.</summary>
/// <remarks>
/// <para>
/// <see cref="Title"/>, <see cref="Source"/> and <see cref="Severity"/> are
/// required. Every other property is optional, and an absent optional property is
/// left out of the payload rather than sent as null.
/// </para>
/// <para>
/// Validation runs when the request is posted, on the calling thread, and is
/// also reachable directly through <see cref="Validate"/>. Object-initializer
/// syntax leaves no constructor to validate in, which is why this type differs
/// from the Java SDK's builder. It is a record so a caller can derive one
/// request from another with <c>with</c>.
/// </para>
/// </remarks>
public sealed record IncidentRequest
{
    /// <summary>The maximum length of <see cref="Source"/>, in characters.</summary>
    public const int MaxSourceLength = 255;

    /// <summary>The incident headline. Required, non-blank.</summary>
    public required string Title { get; init; }

    /// <summary>
    /// The emitting system. Required, non-blank, at most
    /// <see cref="MaxSourceLength"/> characters.
    /// </summary>
    public required string Source { get; init; }

    /// <summary>The severity. Required.</summary>
    public required Severity Severity { get; init; }

    /// <summary>The priority, 1 through 5. Defaults to 1.</summary>
    public int Priority { get; init; } = 1;

    /// <summary>
    /// Milliseconds since the Unix epoch. Defaults to the current time.
    /// </summary>
    public long? SrcTimestamp { get; init; }

    /// <summary>Overrides the service key configured on the client.</summary>
    public string? ServiceKey { get; init; }

    /// <summary>The emitting component.</summary>
    public string? Component { get; init; }

    /// <summary>The owning group.</summary>
    public string? Group { get; init; }

    /// <summary>The incident type.</summary>
    public string? Type { get; init; }

    /// <summary>JSON text sent in the wire-level <c>detailsJSON</c> field.</summary>
    public string? Details { get; init; }

    /// <summary>
    /// Validates the request without posting it. Every post validates anyway, so
    /// this is only useful to reject bad input earlier.
    /// </summary>
    /// <exception cref="ArgumentException">The request is not deliverable.</exception>
    public void Validate() => Normalize();

    /// <summary>
    /// Validates and applies the documented defaults. This is the single place
    /// defaults are resolved, so a synchronous and a queued post can never
    /// disagree about what was sent.
    /// </summary>
    internal NormalizedRequest Normalize()
    {
        RequireNonBlank(Title, nameof(Title));
        RequireNonBlank(Source, nameof(Source));
        if (Source.Length > MaxSourceLength)
        {
            throw new ArgumentException(
                $"{nameof(Source)} must not exceed {MaxSourceLength} characters", nameof(Source));
        }

        if (!Severity.IsDefinedValue())
        {
            throw new ArgumentException($"{nameof(Severity)} must be between 1 and 5", nameof(Severity));
        }

        if (Priority is < 1 or > 5)
        {
            throw new ArgumentException($"{nameof(Priority)} must be between 1 and 5", nameof(Priority));
        }

        var timestamp = SrcTimestamp ?? DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        if (timestamp <= 0)
        {
            throw new ArgumentException(
                $"{nameof(SrcTimestamp)} must be greater than zero", nameof(SrcTimestamp));
        }

        RejectBlank(ServiceKey, nameof(ServiceKey));
        RejectBlank(Component, nameof(Component));
        RejectBlank(Group, nameof(Group));
        RejectBlank(Type, nameof(Type));
        RejectBlank(Details, nameof(Details));

        return new NormalizedRequest(
            Title, Source, (int)Severity, Priority, timestamp, ServiceKey, Component, Group, Type, Details);
    }

    private static void RequireNonBlank(string value, string name)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            throw new ArgumentException($"{name} must not be blank", name);
        }
    }

    /// <summary>
    /// An optional property may be absent, but a present-yet-blank value is
    /// nearly always a bug at the call site, so it is rejected rather than
    /// silently sent.
    /// </summary>
    private static void RejectBlank(string? value, string name)
    {
        if (value is not null && string.IsNullOrWhiteSpace(value))
        {
            throw new ArgumentException($"{name} must not be blank when supplied", name);
        }
    }
}

/// <summary>A validated request with every default resolved.</summary>
internal sealed record NormalizedRequest(
    string Title,
    string Source,
    int Severity,
    int Priority,
    long SrcTimestamp,
    string? ServiceKey,
    string? Component,
    string? Group,
    string? Type,
    string? Details);

File: ruby/lib/oppex_sdk/incident_request.rb
# frozen_string_literal: true

require_relative "errors"
require_relative "severity"

module Oppex
  # A validated, immutable incident submission.
  #
  # Validation happens in the constructor, so a malformed incident is rejected at
  # the call site rather than inside a worker thread where only a log line would
  # remain.
  #
  #   request = Oppex::IncidentRequest.new(
  #     title: "Checkout latency breached the SLO",
  #     source: "checkout-api",
  #     severity: Oppex::Severity::HIGH
  #   )
  #
  # A request may override the service key configured on the client.
  class IncidentRequest
    # The maximum length of +source+, counted in characters.
    MAX_SOURCE_LENGTH = 255

    attr_reader :title, :source, :severity, :priority, :src_timestamp,
                :service_key, :component, :group, :type, :details

    def initialize(title:, source:, severity:, priority: 1, src_timestamp: nil,
                   service_key: nil, component: nil, group: nil, type: nil, details: nil)
      @title = require_non_blank(title, "title")
      @source = require_source(source)
      @severity = Severity.validate!(severity)
      @priority = validate_priority(priority)
      @src_timestamp = validate_timestamp(src_timestamp)
      @service_key = reject_blank(service_key, "service_key")
      @component = reject_blank(component, "component")
      @group = reject_blank(group, "group")
      @type = reject_blank(type, "type")
      @details = reject_blank(details, "details")
      freeze
    end

    private

    def require_non_blank(value, field)
      raise TypeError, "#{field} must be a String" unless value.is_a?(String)
      raise ArgumentError, "#{field} must not be blank" if value.strip.empty?

      value
    end

    def require_source(source)
      source = require_non_blank(source, "source")
      raise ArgumentError, "source must not exceed #{MAX_SOURCE_LENGTH} characters" if source.length > MAX_SOURCE_LENGTH

      source
    end

    # An optional field may be absent, but a present-yet-blank value is nearly
    # always a bug at the call site, so it is rejected rather than silently sent.
    def reject_blank(value, field)
      return nil if value.nil?

      require_non_blank(value, field)
    end

    def validate_priority(priority)
      raise TypeError, "priority must be an Integer" unless priority.is_a?(Integer)
      raise ArgumentError, "priority must be between 1 and 5" unless (1..5).cover?(priority)

      priority
    end

    def validate_timestamp(src_timestamp)
      return (Time.now.to_f * 1000).to_i if src_timestamp.nil?

      raise TypeError, "src_timestamp must be an Integer" unless src_timestamp.is_a?(Integer)
      raise ArgumentError, "src_timestamp must be greater than zero" unless src_timestamp.positive?

      src_timestamp
    end
  end
end
